tempesta_fw/http.c

```diff
@@ -6092,69 +6103,65 @@ __tfw_http_msg_body_dup(const char *filename, TfwStr *c_len_hdr, size_t *len,
 			size_t *body_offset)
 {
 	char *body, *b_start, *res = NULL;
-	size_t b_sz, t_sz;
+	size_t b_sz, t_sz = 0;
 	char buff[TFW_ULTOA_BUF_SIZ] = {0};
-	TfwStr *cl_buf = __TFW_STR_CH(c_len_hdr, 1);
+	TfwStr *cl_buf = c_len_hdr ? __TFW_STR_CH(c_len_hdr, 1) : 0;
 
 	body = tfw_cfg_read_file(filename, &b_sz, 0);
 	if (!body) {
 		*len = *body_offset = 0;
 		return NULL;
 	}
-	cl_buf->data = buff;
-	cl_buf->len = tfw_ultoa(b_sz, cl_buf->data, TFW_ULTOA_BUF_SIZ);
-	if (unlikely(!cl_buf->len)) {
-		T_ERR_NL("Can't copy file %s: too big\n", filename);
-		goto err;
+	if (c_len_hdr) {
+		cl_buf->data = buff;
+		cl_buf->len = tfw_ultoa(b_sz, cl_buf->data, TFW_ULTOA_BUF_SIZ);
+		if (unlikely(!cl_buf->len)) {
+			T_ERR_NL("Can't copy file %s: too big\n", filename);
+			goto err;
+		}
+
+		c_len_hdr->len += cl_buf->len;
+		t_sz += c_len_hdr->len;
 	}
 
-	c_len_hdr->len += cl_buf->len;
-	t_sz = c_len_hdr->len + b_sz;
-	res = (char *)__get_free_pages(GFP_KERNEL, get_order(t_sz));
+	t_sz += b_sz;
+	b_start = res = (char *)__get_free_pages(GFP_KERNEL, get_order(t_sz));
 	if (!res) {
 		T_ERR_NL("Can't allocate memory storing file %s as response "
 			 "body\n", filename);
 		goto err_2;
 	}
 
-	tfw_str_to_cstr(c_len_hdr, res, t_sz);
-	b_start = res + c_len_hdr->len;
+	if (c_len_hdr) {
+		tfw_str_to_cstr(c_len_hdr, res, t_sz);
+		b_start += c_len_hdr->len;
+	}
 	memcpy_fast(b_start, body, b_sz);
 
 	*len = t_sz;
 	*body_offset = b_start - res;
 err_2:
-	c_len_hdr->len -= cl_buf->len;
+	if (c_len_hdr)
+		c_len_hdr->len -= cl_buf->len;
 err:
-	cl_buf->data = NULL;
-	cl_buf->len = 0;
 	free_pages((unsigned long)body, get_order(b_sz));
 
 	return res;
 }
 
 /**
- * Copy @filename content to allocated memory as compound of
- * `Content-length' header, crlfcrlf and message body. Memory is allocated
- * via __get_free_pages(), thus free_pages() must be used on cleanup;
+ * Copy @filename content to allocated memory. Memory is allocated
+ * via __get_free_pages(), thus free_pages() must be used on cleanup.
+ * Unlike in @__tfw_http_msg_body_dup() content-length header is not added
+ * to the body.
  * @len		- total length of body data including headers.
  */
 char *
 tfw_http_msg_body_dup(const char *filename, size_t *len)
 {
-	TfwStr c_len_hdr = {
-		.chunks = (TfwStr []){
-			{ .data = S_F_CONTENT_LENGTH,
-			  .len = SLEN(S_F_CONTENT_LENGTH) },
-			{ .data = NULL, .len = 0 },
-			{ .data = S_CRLFCRLF, .len = SLEN(S_CRLFCRLF) },
-		},
-		.len = SLEN(S_F_CONTENT_LENGTH S_CRLFCRLF),
-		.nchunks = 3
-	};
 	size_t b_off;
 
-	return __tfw_http_msg_body_dup(filename, &c_len_hdr, len, &b_off);
+	return __tfw_http_msg_body_dup(filename, NULL, len, &b_off);
```

Comment:
`__tfw_http_msg_body_dup()` is called only in http.c, so make it static

Reply:
Fixed.